base/combinatorics.jl

```diff
@@ -59,29 +66,31 @@ julia> isperm([1; 3])
 false
 ```
 """
-isperm(A) = _isperm(A)
-
-function isperm(P::NTuple{N,Int}) where N
-    if N>21 # if N above 21 other algorithm is faster
-        return _isperm(P)
-    end
-    for i in eachindex(P)
-        flag=true
-        for j in eachindex(P)
-            if P[j]==i
-                flag=false
-                break
-            end
-        end
-        flag && return false
+function isperm(A)
+    n = length(A)
+    used = falses(n)
+    for a in A
+        (0 < a <= n) && (used[a] ⊻= true) || return false
     end
-    return true
+    true
 end
 
 isperm(p::Tuple{}) = true
 isperm(p::Tuple{Int}) = p[1] == 1
 isperm(p::Tuple{Int,Int}) = ((p[1] == 1) & (p[2] == 2)) | ((p[1] == 2) & (p[2] == 1))
 
+function isperm(P::NTuple{N,Integer}) where {N}
+    valn = Val(N)
+    _foldoneto(true, valn) do b,i
+        s = _foldoneto(false, valn) do s, j
+            s && return s
+            P[j]==i && return true
+            false
```

Comment:
I think you can simplify this to

```suggestion
            s || P[j]==i
```